actors/verifreg/src/lib.rs

```diff
@@ -528,6 +521,87 @@ impl Actor {
         Ok(GetClaimsReturn { batch_info: batch_gen.gen(), claims })
     }
 
+    /// Extends the maximum term of some claims up to the largest value they could have been
+    /// originally allocated.
+    /// Callable only by the claims' client.
+    /// Cannot reduce a claim's term.
+    /// Can extend the term even if the claim has already expired.
+    /// Note that this method can't extend the term past the original limit,
+    /// even if the term has previously been extended past that by spending new datacap.
+    pub fn extend_claim_terms<BS, RT>(
+        rt: &mut RT,
+        params: ExtendClaimTermsParams,
+    ) -> Result<ExtendClaimTermsReturn, ActorError>
+    where
+        BS: Blockstore,
+        RT: Runtime<BS>,
+    {
+        // Permissions are checked per-claim.
+        rt.validate_immediate_caller_accept_any()?;
+        let caller_id = rt.message().caller().id().unwrap();
+        let term_limit = rt.policy().maximum_verified_allocation_term;
+
+        let mut batch_gen = BatchReturnGen::new(params.terms.len());
+        rt.transaction(|st: &mut State, rt| {
+            let mut st_claims = st.load_claims(rt.store())?;
+            // Group consecutive term extensions with the same provider so we can batch update
```

Comment:
Were you making this decision keeping in mind mapmap caching?  I agree this saves checking the mapmap's cache which is something but this points to redundancy.  If the mapmap's cache handling was slow enough that you didn't want to rely on it then 1) we should remove that complexity from the data type and accept caller level complexity like this, or 2) we should improve mapmap caching to be useful.

Reply:
I did not have mapmap caching specifically in mind. My motivation was somewhat abstract toward the idea that batch updating a H/AMT should be more efficient due to only calculating intermediate nodes once, even if many values are on the same path. I am aware that the underlying HAMT doesn't actually implement this yet, but felt compelled to do this in the "right" way anyway, so it would immediately benefit if that lower-level improvement is made.

It would be possible to improve either `MapMap`, or just the `Map` or HAMT, to cache all updates in memory and only serialize once at the end. That would be more extensive and tricky caching than the present approach, but remove all such considerations and complexity from caller-level code like this. I'm hesitant to push on it due to difficulty of getting it right at that layer.

Reply:
> cache all updates in memory and only serialize once at the end

Doesn't internal HAMT node caching solve this problem already?  I think it does and I don't think you need the caller complexity added here to achieve that.  You already get this with HAMT + MapMap caching.

Reply:
Ok. If the HAMT does do that, it's not very obvious or at all documented. I tried to quickly benchmark this to see, but discovered much pain trying to replace the mock runtime's blockstore with a TrackingBlockstore (which comes from a different repo (😠 see #678)). I'll revert it on the principle of simplicity first.

Reply:
Not surprised if the rust library has no mention of this but this used to be a consensus thing before fvm so this was documented here: https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0007.md